scripts/helper_scripts/unipept-database-rs/src/calculate_lcas/taxonomy.rs

```diff
@@ -0,0 +1,161 @@
+use std::collections::HashMap;
+use std::io::BufRead;
+use std::path::PathBuf;
+
+use anyhow::{Context, Result};
+
+use crate::taxons_uniprots_tables::utils::now_str;
+use crate::utils::files::{open_read, open_sin};
+
+const GENUS: u8 = 18;
+const RANKS: u8 = 27;
+const SPECIES: u8 = 22;
+const NULL: &str = "\\N";
+const SEPARATOR: &str = "\t";
+
+pub struct Taxonomy {
+    taxonomy: Vec<Vec<i32>>,
+}
+
+impl Taxonomy {
+    pub fn build(infile: &PathBuf) -> Result<Self> {
+        let mut taxonomy_map: HashMap<i32, Vec<i32>> = HashMap::new();
+        let reader = open_read(infile).context("Unable to open input file")?;
+
+        let mut max = i32::MIN;
+
+        for line in reader.lines() {
+            let line = line.with_context(|| {
+                format!("Error reading line from input file {}", infile.display())
+            })?;
+            let elements: Vec<String> = line.splitn(28, SEPARATOR).map(String::from).collect();
+
+            let key = parse_int(&elements[0])?;
+            // Note on the collect::<> here: "?" can't be used inside of map() as it is a closure
+            // Collecting into a Result<Vec<_>> will stop instantly when it receives one Error
+            // https://doc.rust-lang.org/rust-by-example/error/iter_result.html#fail-the-entire-operation-with-collect
+            let lineage = elements
+                .iter()
+                .skip(1)
+                .map(parse_int)
+                .collect::<Result<Vec<i32>>>()?;
+            taxonomy_map.insert(key, lineage);
```

Comment:
Why not

```suggestion
            let mut elements = line.splitn(28, SEPARATOR).map(parse_int);
            let key = elements.next().unwrap()?;
            // Note on the collect::<> here: "?" can't be used inside of map() as it is a closure
            // Collecting into a Result<Vec<_>> will stop instantly when it receives one Error
            // https://doc.rust-lang.org/rust-by-example/error/iter_result.html#fail-the-entire-operation-with-collect
            let lineage = elements.collect::<Result<Vec<i32>>>()?;
            taxonomy_map.insert(key, lineage);
```

I think this is somewhat cleaner and avoids allocating a `String` I think. This requires changing `parse_int` to take a `&str` instead of a `&String` though,